File: medium/cpp/2265.cpp
// https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    int averageOfSubtree(TreeNode* root)
    {
        int answer = 0;
        
        return std::get<2>(solve(root));
    }
    
private:
     std::tuple<int, int, int> solve(TreeNode* root)
     {
         if (root == nullptr)
         {
             return { 0, 0, 0 };
         }
         
         const auto left = solve(root->left);
         const auto right = solve(root->right);
         
         int tot = 1 + std::get<0>(left) + std::get<0>(right);
         int sum = root->val + std::get<1>(left) + std::get<1>(right);
         int answer = std::get<2>(left) + std::get<2>(right);
         
         if (sum / tot == root->val)
         {
             answer++;
         }
         
         return { tot, sum, answer };
     }
};

File: medium/python/2265.py
# https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        cnt = [0]
        self.solve(root, cnt)
        return cnt[0]

    def solve(self, root, cnt):
        if root is None:
            return (0, 0)

        cur = tuple(sum(x) for x in zip((1, root.val), self.solve(root.left, cnt), self.solve(root.right, cnt)))

        if root.val == cur[1] // cur[0]:
            cnt[0] += 1

        return cur